### PR TITLE
add Utilities::MPI::sum() for FullMatrix objects

### PR DESCRIPTION
**Pull request by davydden**

### CODE REVIEW

**Comment by drwells** [comment]
Would you please add a test for a complex number type in addition to float and double?

My other criticism is meta and if you aren't willing to do this then we can just open an issue later. We currently have six versions of `MPI::sum` and this adds a seventh. They could all be replaced by

```cpp
template <typename Number>
void
sum (const ArrayView<const Number> &local,
     const MPI_Comm                &communicator,
     ArrayView<Number>              result);
```

and (for convenience) a corresponding scalar version that just creates a one-entry `ArrayView` and forwards everything to the general one. The `all_reduce`, `min`, and `max` functions could similarly be adapted to use `ArrayView`s (right now they do not support as many container types as `sum`).

**Comment by davydden** [comment]
> Would you please add a test for a complex number type in addition to float and double?

done. It's WIP as for whatever reasons it times-out with 10 cores....

 >  we can just open an issue later

I would prefer a separate issue for the refactoring you suggest.

File: tests/mpi/collective_full_matrix.cc [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+      test<double>();
+      deallog.pop();
+      deallog.push("complex<double>");
+      test<std::complex<double>>();
+      deallog.pop();
+    }
+  else
+    {
+      test<float>();
+      test<double>();
```
Comment (drwells) [inline tests/mpi/collective_full_matrix.cc]: Shouldn't there be a call to `test<std::complex<double>>` here?

Comment (davydden) [reply tests/mpi/collective_full_matrix.cc]: yes-yes-yes, thanks, that's why it's not working :smile:

**Comment by drwells** [comment]
LGTM

Since this is a MPI test lets run the serial tester too.

/run-tests

**Pull request merged**